File: demo/freshness.py
# demo/freshness.py
"""Does the connected index still match the repository?

A connected repo's corpus is frozen at the commit it was ingested. Nothing
told anyone that. This repo's own index sat NINE COMMITS behind HEAD while
answering questions with complete confidence, and the only reason it was
noticed was that someone happened to add a file and go looking for it.

That is the same failure this product exists to avoid, moved out of citations
and into time: every answer was grounded in evidence genuinely retrieved from
a corpus that no longer described the repository.

## The one property everything here is arranged around

**"I could not check" must never render as "up to date".** `up_to_date` is
three-valued -- True, False, or **None for unknown** -- and every failure path
lands on None. A network blip, a revoked token, a rate limit: all unknown,
none of them reassuring. Telling someone their index is current when we did
not manage to look is a confident claim the evidence does not support, which
is the thing this codebase refuses to do everywhere else.

One deliberate exception, in the other direction: if HEAD is readable and
DIFFERS from the indexed commit but the comparison call then fails,
`up_to_date` stays **False** and only `behind_by` is None. We genuinely know
it differs; downgrading that to unknown would discard a fact we hold.

## Why it is cached

`/status` is polled continuously by the Mac app, and GitHub's API is not. The
result is cached per (repo, indexed commit) for a TTL, and `checked_at`
reports when the check ACTUALLY ran -- not now -- so a reader knows the
verdict may be minutes old. Keying on the indexed commit as well as the repo
means a refresh invalidates the entry immediately: otherwise the moment a user
finishes a refresh and looks at the banner to watch it clear is exactly the
moment they would be shown the pre-refresh verdict.

The caller's token is used for the request and never stored -- a checker is
shared across a repo's users, so a retained token would be one caller's
credential spent on another's request.
"""

import threading

from evals import github_access

_DEFAULT_TTL = 600.0  # seconds; a repo does not drift meaningfully faster


def _unknown(checked_at=None):
    """Every key present on every path, so a client can never KeyError its way
    into rendering nothing."""
    return {"up_to_date": None, "behind_by": None,
            "head_commit": None, "checked_at": checked_at}


class FreshnessChecker:
    """Thread-safe, TTL-cached staleness lookups for connected repos."""

    def __init__(self, head_fn=None, between_fn=None, now=None, ttl=_DEFAULT_TTL):
        self._head = head_fn or (lambda repo, token: github_access.head_commit(repo, token))
        self._between = between_fn or (
            lambda repo, base, target, token: github_access.commits_between(repo, base, target, token))
        if now is None:
            import time
            now = time.time
        self._now = now
        self._ttl = ttl
        self._cache = {}
        self._lock = threading.Lock()

    def check(self, repo, indexed_commit, token):
        """-> {up_to_date, behind_by, head_commit, checked_at}. Never raises.

        `token` is used for this request only and is never written to the
        cache, which is keyed on (repo, indexed_commit) alone.
        """
        if not repo or not indexed_commit:
            return _unknown()

        key = (repo, indexed_commit)
        with self._lock:
            cached = self._cache.get(key)
            if cached is not None and self._now() - cached["checked_at"] < self._ttl:
                return dict(cached)

        checked_at = self._now()
        head = self._head(repo, token)
        if not head:
            # Unknown: NOT reported as up to date, and deliberately NOT cached
            # either -- a failed check should be retried on the next poll
            # rather than pinned for the whole TTL.
            return _unknown(checked_at)

        if head == indexed_commit:
            result = {"up_to_date": True, "behind_by": 0,
                      "head_commit": head, "checked_at": checked_at}
        else:
            behind = self._between(repo, indexed_commit, head, token)
            result = {"up_to_date": False, "behind_by": behind,
                      "head_commit": head, "checked_at": checked_at}

        with self._lock:
            self._cache[key] = dict(result)
        return result

File: mac/Icarus/Sources/IcarusKit/Freshness.swift
import Foundation

/// Does the connected index still match the repository? (`demo/freshness.py`)
///
/// A connected corpus is frozen at the commit it was ingested. Nothing used to
/// say so, and this repo's own index once sat nine commits behind HEAD while
/// answering with complete confidence — the product's own failure mode moved
/// out of citations and into time.
///
/// **The one rule: "I could not check" must never render as "up to date".**
/// The wire format makes `up_to_date` three-valued for exactly that reason,
/// and this type keeps it three-valued all the way to the view via
/// `IndexFreshness` — see that enum for why a `Bool?` would not have been good
/// enough.
public struct Freshness: Decodable, Equatable, Sendable {
    /// True, false, or **nil for unknown**. Read `state`, not this, unless you
    /// have a reason to care about the raw wire value.
    public let upToDate: Bool?
    /// How many commits the repository has moved since the indexed commit.
    /// nil means the count is unavailable — which is NOT the same as zero.
    public let behindBy: Int?
    /// The built-in demo corpus is frozen deliberately (it is the reproducible
    /// eval board), so it is permanently behind upstream. Absent reads as
    /// false, which is correct for every real connected repo.
    public let pinned: Bool?

    private enum CodingKeys: String, CodingKey {
        case pinned
        case upToDate = "up_to_date"
        case behindBy = "behind_by"
    }

    public init(upToDate: Bool?, behindBy: Int?, pinned: Bool? = nil) {
        self.upToDate = upToDate
        self.behindBy = behindBy
        self.pinned = pinned
    }

    public var state: IndexFreshness {
        if pinned == true { return .pinned(behindBy) }
        switch upToDate {
        case true: return .matches
        case false: return .behind(behindBy)
        default: return .unknown
        }
    }
}

/// Index staleness as a closed set of cases.
///
/// Deliberately an enum rather than the `Bool?` it decodes from. An optional
/// Bool invites `?? false` at the call site, which would render "I could not
/// check" as "up to date" — a confident claim the evidence does not support,
/// and the same class of failure as a bluffed citation. With an enum there is
/// no default to fall into: every view must name `.unknown` explicitly.
public enum IndexFreshness: Equatable, Sendable {
    /// The index is at the repository's current head.
    case matches
    /// The index is behind. The count is nil when it could not be determined —
    /// we know it differs, we just cannot say by how much, and that remains
    /// more honest than pretending it matches.
    case behind(Int?)
    /// The check did not complete. NOT "up to date", NOT an error — a third
    /// state, and it must look like one.
    case unknown
    /// The built-in demo corpus, frozen on purpose and not refreshable.
    /// Reporting its real distance from upstream as ordinary staleness would
    /// describe a deliberate decision as neglect.
    case pinned(Int?)

    /// One line a user can act on. Never claims currency it has not verified.
    public var summary: String {
        switch self {
        case .matches:
            return "Index matches the repository."
        case .behind(let n?):
            return "\(n) commit\(n == 1 ? "" : "s") behind the repository."
        case .behind(nil):
            return "Behind the repository — Icarus couldn’t work out by how much."
        case .unknown:
            return "Icarus couldn’t check whether this index is current."
        case .pinned(let n?):
            return "Built-in demo corpus, frozen on purpose — \(n) commits behind upstream."
        case .pinned(nil):
            return "Built-in demo corpus, frozen on purpose."
        }
    }

    /// Should a client offer to refresh? Never for the pinned demo corpus,
    /// whose refresh is forbidden by design and would do nothing.
    public var offersRefresh: Bool {
        switch self {
        case .behind: return true
        case .matches, .unknown, .pinned: return false
        }
    }
}
